### PR TITLE
fix(bpf): collect close/detach errors and add nil checks in BPF manager

### PR DESCRIPTION
**Pull request by wc4440222**
## What does this PR do?

Improves error handling and safety in the BPF manager (`internal/bpf/bpf_default.go`).

### 1. `Close()` — Collect errors instead of always returning nil

**Before**: `Close()` discarded all individual map/program/link close errors and always returned `nil`, making it impossible for callers to detect cleanup failures.

**After**: Collects all close errors and returns them via `errors.Join`. Each error includes the map/program/link name for debuggability.

### 2. `Detach()` — Collect and return errors instead of silently discarding

**Before**: `Detach()` only logged errors via `log.Debugf` and always returned `nil`.

**After**: Collects link close and perf event detach errors and returns them via `errors.Join`.

### 3. `AttachWithOptions()` and `Attach()` — Log detach errors during failure cleanup

**Before**: `_ = b.Detach()` silently discarded errors when cleaning up after an attach failure.

**After**: Logs detach errors with `log.Warnf` so failures are visible.

### 4. Nil checks to prevent panics on partial initialization

**Before**: `Close()` and `Detach()` could panic if `bMap`, `bProg`, or link entries were nil (e.g., due to partial initialization failure).

**After**: All entries are nil-checked before calling `Close()`.

## Testing

- `go build ./internal/bpf/` — compiles successfully ✅
- Existing `bpf_default_test.go` tests pass (the `TestManager_InitAndClose` test requires root for eBPF loading, which is expected)

## Checklist

- [x] Code follows the project's style guidelines
- [x] Self-reviewed the code
- [x] No breaking API changes (function signatures unchanged)
- [x] DCO signed (`Signed-off-by`)

Signed-off-by: wc4440222 <199748891+wc4440222@users.noreply.github.com>

### CODE REVIEW

**Comment by huhong-web** [comment]
## Code Review: fix(bpf): collect close/detach errors and add nil checks in BPF manager

**PR**: #307 | **Author**: @wc4440222 | **File**: `internal/bpf/bpf_default.go` (+48/-14)

---

### 1. Copyright Check
- **PASS** — No new files added. The modified file already has a valid Apache 2.0 Copyright header (2025).

### 2. Golang Style Check
- **PASS** — Error wrapping with `%w` is correct. `errors.Join` usage is idiomatic Go 1.20+. Variable declarations follow conventions. Nil slice initialization via `var errs []error` is appropriate for `append` + `errors.Join` pattern.

### 3. Abstract Interface Check
- **PASS** — The `BPF` interface in `bpf.go` is unchanged. Function signatures (`Close() error`, `Detach() error`) remain compatible. No new abstractions needed.

### 4. Architecture Check
- **PASS** — SRP: `Close()` handles full cleanup, `Detach()` handles link removal only — proper separation. SOLID principles are respected.

### 5. Security Check
- **PASS** — No injection, hardcoded secrets, path traversal, or crypto concerns. Error messages expose internal BPF map/program/link names, which are system identifiers — no information disclosure risk.

---

### Issues Found

#### P1: `Close()` is not idempotent — double-close via `runtime.SetFinalizer` will return errors

The constructor sets `runtime.SetFinalizer(b, (*defaultBPF).Close)`. If `Close()` is called explicitly, the GC finalizer will call it again, attempting to close already-freed BPF resources (fds). With the old code this silently returned `nil`; now it will return `errors.Join` of EBADF/close errors, which is noisy and confusing.

**Fix**: Clear the finalizer at the start of `Close()` and set `innerPerfEvent = nil` after detaching:

```go
func (b *defaultBPF) Close() error {
    runtime.SetFinalizer(b, nil)  // prevent double-close via GC
    var closeErrs []error
    // ... existing close logic ...

    if b.innerPerfEvent != nil {
        if err := b.innerPerfEvent.detach(); err != nil {
            closeErrs = append(closeErrs, fmt.Errorf("detach perf event: %w", err))
        }
        b.innerPerfEvent = nil
    }
    return errors.Join(closeErrs...)
}
```

#### P2: `Detach()` does not clear closed links from `spec.links` map — re-attach will fail

After closing links in `Detach()`, the map entries are not removed. The attach methods (`attachKprobe`, `attachTracepoint`, `attachRawTracepoint`) check for duplicate `linkKey` and return `"duplicate symbol"` errors. So a `Detach()` → `Attach()` cycle will fail unexpectedly.

**Fix**: Delete each link from the map after closing it:

```go
for linkKey, l := range spec.links {
    if l != nil {
        if err := l.Close(); err != nil {
            detachErrs = append(detachErrs, fmt.Errorf("detach link %s in program %s: %w", linkKey, spec.name, err))
        }
    }
    delete(spec.links, linkKey)
}
```

Also set `b.innerPerfEvent = nil` after detaching to prevent double-detach if `Close()` is called after `Detach()`.

#### P3: `Close()` and `Detach()` both detach `innerPerfEvent` — double-detach if both are called

If `Detach()` is called first and then `Close()` is called, `innerPerfEvent.detach()` will be invoked twice. The original code had the same issue but silently discarded errors; now it will surface a spurious error.

**Fix**: Set `b.innerPerfEvent = nil` after detaching in both methods (as shown in P1 and P2 fixes above).

---

### Summary

| Category | Verdict |
|----------|---------|
| Copyright | ✅ PASS |
| Go Style | ✅ PASS |
| Interface | ✅ PASS |
| Architecture | ✅ PASS |
| Security | ✅ PASS |
| **Overall** | ⚠️ **needs-changes** (3 issues, P1-P3) |

The error collection and nil-check improvements are solid and well-structured. The main gap is **idempotency**: without clearing the finalizer, link map entries, and `innerPerfEvent` reference after cleanup, the improved error reporting will surface confusing errors on double-close/detach paths that previously silently passed.

**Comment by hao022** [comment]
@wc4440222 因为代码提交出现了冲突，我将你的commit1合入了主线，commit2 没有采用；最终的补丁链接为：https://github.com/ccfos/huatuo/commit/7a17652117420ac9b08d19f733ba4205a9aa7c10 感谢你的贡献！
